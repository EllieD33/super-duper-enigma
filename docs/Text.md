# Text Component

## Description

The `Text` component provides flexible text rendering with customisable styles. It allows you to control the tag, typography (weight, size, family), colour, and alignment of the text. This component is suitable for various text elements, such as headings, paragraphs, and other content types.

## Props

### `TextProps`

| Prop        | Type                            | Required | Description                                                           |
| ----------- | ------------------------------- | -------- | --------------------------------------------------------------------- |
| `children`  | `React.ReactNode`               | ✅       | The content to display inside the text component.                     |
| `as`        | `TextTag`                       | ❌       | Specifies the HTML tag to render (e.g., `h1`, `p`). Defaults to `p`.  |
| `styleAs`   | `keyof typeof TextTag`          | ❌       | Defines the style variant to apply (overrides `as` prop if provided). |
| `colour`    | `Colours`                       | ❌       | The text colour (from custom `Colours` constant). Defaults to `Gray`. |
| `weight`    | `FontWeight`                    | ❌       | The font weight (from custom `FontWeight` constant).                  |
| `size`      | `FontSize`                      | ❌       | The font size (from custom `FontSize` constant).                      |
| `font`      | `FontFamily`                    | ❌       | The font family (from custom `FontFamily` constant).                  |
| `align`     | `"left" \| "centre" \| "right"` | ❌       | Aligns the text (default is `left`).                                  |
| `className` | `string`                        | ❌       | Additional custom class names to apply to the text element.           |

## Behaviour

- The component dynamically renders the specified `as` tag (e.g., `h1`, `p`, etc.) with the provided typography styles.
- The `styleAs` prop allows you to apply predefined styles based on the text tag (such as `Heading1`, `Paragraph`).
- The `align` prop controls text alignment (e.g., left, center, or right).
- Custom styles, such as `colour`, `weight`, `size`, and `font`, are applied directly to the inline style of the text.
- A warning is displayed in the console if the `children` are not valid for the provided `as` tag.

[Explore Text in Storybook](http://localhost:6006/?path=/story/library-text-text--default&globals=viewport:largeTablet)

## Notes

- The component uses `clsx` to conditionally apply CSS class names, allowing for additional styling flexibility.
- The `Typography` object from constants is used to provide default styles based on the tag or style variant.
- The `isValidTagForChildren` helper function ensures that the children are compatible with the selected tag (e.g., not rendering block elements within inline tags).
