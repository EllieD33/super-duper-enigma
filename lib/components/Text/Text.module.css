* {
    box-sizing: border-box;
    margin: 0;
    padding: 0;
}

.left {
    text-align: left;
}

.centre {
    text-align: center;
}

.right {
    text-align: right;
}
