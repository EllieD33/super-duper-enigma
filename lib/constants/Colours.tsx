export enum Colours {
    White = "#f9f9f9",
    Black = "#000300",
    Gray = "#333333",

    Red = "#EF5350",
    Blue = "#2C699A",
    DarkBlue = "#073B4C",
    Green = "#16DB93",
    Yellow = "#EFEA5A",
    Purple = "#54478C",
    Orange = "#F29E4C",
    Teal = "#0DB39E",
    Pink = "#EF476F",
}

export enum PastelColors {
    Orange = "#F2BC8E",
    Blue = "#A1BFD7",
    Red = "#F1A8A0",
    Green = "#A8E6A3",
    Purple = "#9E8DC0",
    Yellow = "#F2F2A9",
    Gray = "#B0B0B0",
    Teal = "#8ED1C5",
    Pink = "#F7A8B8",
}

export enum ErrorColours {
    Light = "#EF9A9A",
    Main = "#F44336",
    Dark = "#D32F2F",
}

export enum WarningColours {
    Light = "#FFE082",
    Main = "#FFC107",
    Dark = "#FFA000",
}

export enum SuccessColours {
    Light = "#A5D6A7",
    Main = "#4CAF50",
    Dark = "#388E3C",
}

export enum InfoColours {
    Light = "#90CAF9",
    Main = "#2196F3",
    Dark = "#1565C0",
}
