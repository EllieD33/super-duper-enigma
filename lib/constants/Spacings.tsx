export enum Spacing {
    Spacing0 = 0,
    Spacing1 = 4,
    Spacing2 = 8,
    Spacing3 = 12,
    Spacing4 = 16,
    Spacing5 = 24,
    Spacing6 = 32,
    Spacing7 = 40,
    Spacing8 = 48,
    Spacing9 = 64,
    Spacing10 = 80,
    Spacing11 = 96,
    Spacing12 = 128,
}
