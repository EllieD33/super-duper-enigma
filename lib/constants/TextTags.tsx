export enum TextTag {
    Span = "span",
    Paragraph = "p",
    Heading1 = "h1",
    Heading2 = "h2",
    Heading3 = "h3",
    Heading4 = "h4",
    Heading5 = "h5",
    Heading6 = "h6",
    Strong = "strong",
    Emphasis = "em",
    Label = "label",
    Small = "small",
    Div = "div",
}
