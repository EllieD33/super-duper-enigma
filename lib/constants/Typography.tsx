export enum FontWeight {
    Thin = 100,
    ExtraLight = 200,
    Light = 300,
    Regular = 400,
    Medium = 500,
    SemiBold = 600,
    Bold = 700,
    ExtraBold = 800,
    Black = 900,
}

export enum TextTransform {
    None = "none",
    Uppercase = "uppercase",
    Lowercase = "lowercase",
    Capitalise = "capitalise",
}

export enum FontSize {
    Small = "0.8rem",
    Medium = "1rem",
    Large = "1.25rem",
    XL = "1.5rem",
    XXL = "2rem",
    XXXL = "3rem",
    Heading1 = "2rem",
    Heading2 = "1.75rem",
    Heading3 = "1.5rem",
    Heading4 = "1.25rem",
    Heading5 = "1.125rem",
    Heading6 = "1rem",
}

export enum FontFamily {
    Heading = "'Roboto', sans-serif",
    Body = "'Open Sans', sans-serif",
    System = "system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Helvetica Neue', Arial, sans-serif",
}

export const Typography = {
    Heading1: {
        fontSize: FontSize.Heading1,
        fontWeight: FontWeight.Bold,
        fontFamily: FontFamily.Heading,
    },
    Heading2: {
        fontSize: FontSize.Heading2,
        fontWeight: FontWeight.Bold,
        fontFamily: FontFamily.Heading,
    },
    Heading3: {
        fontSize: FontSize.Heading3,
        fontWeight: FontWeight.Bold,
        fontFamily: FontFamily.Heading,
    },
    Heading4: {
        fontSize: FontSize.Heading4,
        fontWeight: FontWeight.Bold,
        fontFamily: FontFamily.Heading,
    },
    Heading5: {
        fontSize: FontSize.Heading5,
        fontWeight: FontWeight.Bold,
        fontFamily: FontFamily.Heading,
    },
    Heading6: {
        fontSize: FontSize.Heading6,
        fontWeight: FontWeight.Bold,
        fontFamily: FontFamily.Heading,
    },
    Paragraph: {
        fontSize: FontSize.Medium,
        fontWeight: FontWeight.Regular,
        fontFamily: FontFamily.Body,
    },
    Span: {
        fontSize: FontSize.Medium,
        fontWeight: FontWeight.Regular,
        fontFamily: FontFamily.Body,
    },
    Strong: {
        fontSize: FontSize.Medium,
        fontWeight: FontWeight.Bold,
        fontFamily: FontFamily.Body,
    },
    Emphasis: {
        fontSize: FontSize.Medium,
        fontWeight: FontWeight.Regular,
        fontFamily: FontFamily.Body,
        fontStyle: "italic",
    },
    Label: {
        fontSize: FontSize.Small,
        fontWeight: FontWeight.Medium,
        fontFamily: FontFamily.Body,
    },
    Small: {
        fontSize: FontSize.Small,
        fontWeight: FontWeight.Regular,
        fontFamily: FontFamily.Body,
    },
    Div: {
        fontSize: FontSize.Medium,
        fontWeight: FontWeight.Regular,
        fontFamily: FontFamily.Body,
    },
} as const;
