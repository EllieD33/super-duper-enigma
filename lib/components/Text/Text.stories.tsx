import type { Meta, StoryObj } from "@storybook/react";
import Text, { TextProps } from "./Text";
import { TextTag } from "../../constants/TextTags";
import { Colours } from "../../constants/Colours";
import { FontFamily } from "../../constants/Typography";

const meta = {
    title: "Library/Text/Text",
    component: Text,
    parameters: {
        layout: "fullscreen",
    },
    tags: ["autodocs"],
    argTypes: {},
} satisfies Meta<typeof Text>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        children: "This is what default text looks like",
    } as TextProps,
};

export const Heading: Story = {
    args: {
        children: "This is what a default heading looks like",
        as: TextTag.Heading1,
    } as TextProps,
};
