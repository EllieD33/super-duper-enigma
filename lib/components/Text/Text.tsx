import React, { ReactElement } from "react";
import { TextTag } from "../../constants/TextTags";
import { Colours } from "../../constants/Colours";
import { FontFamily, FontSize, FontWeight } from "../../constants/Typography";
import styles from "./Text.module.css";

export interface TextProps {
    children: React.ReactNode;
    as?: TextTag;
    colour?: Colours;
    weight?: FontWeight;
    size?: FontSize;
    font?: FontFamily;
    align?: "left" | "centre" | "right";
}

const Text = ({
    children,
    as: Tag = TextTag.Paragraph,
    colour = Colours.Gray,
    weight = FontWeight.Regular,
    size = FontSize.Medium,
    font = FontFamily.Body,
    align = "left",
}: TextProps): ReactElement => {
    const inlineStyles = {
        color: colour,
        fontWeight: Tag.startsWith("h") ? FontWeight.Bold : weight,
        fontSize: Tag.startsWith("h")
            ? FontSize[`${Tag}` as keyof typeof FontSize]
            : size,
        fontFamily: Tag.startsWith("h") ? FontFamily.Heading : font,
    };

    return (
        <Tag className={styles[align as keyof TextProps]} style={inlineStyles}>
            {children}
        </Tag>
    );
};

export default Text;
