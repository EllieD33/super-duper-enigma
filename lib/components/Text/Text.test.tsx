import React from "react";
import { render, screen } from "@testing-library/react";
import Text, { TextProps } from "./Text";
import { TextTag } from "../../constants/TextTags";

const defaultProps: TextProps = {
    children: "I am some test text",
};

const headingTags = [
    TextTag.Heading1,
    TextTag.Heading2,
    TextTag.Heading3,
    TextTag.Heading4,
    TextTag.Heading5,
    TextTag.Heading6,
];

describe("Text", () => {
    it("should render the default component", () => {
        const { container } = render(<Text>{defaultProps.children}</Text>);
        expect(container.firstChild).toMatchSnapshot();
    });

    headingTags.forEach((heading) => {
        it(`"should apply Heading font family to ${heading} tag`, () => {
            const { container } = render(
                <Text as={heading}>{defaultProps.children}</Text>
            );
            expect(container.firstChild).toHaveStyle(
                "font-family: 'Roboto', sans-serif"
            );
        });
    });
});
