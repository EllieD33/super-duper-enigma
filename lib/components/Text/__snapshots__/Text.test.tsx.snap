// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should render the default component 1`] = `
<p
  class="left"
  style="color: rgb(51, 51, 51); font-weight: 400; font-size: 1rem; font-family: 'Open Sans', sans-serif;"
>
  I am some test text
</p>
`;
